crates/oxc_linter/src/rules/typescript/no_unsafe_declaration_merging.rs

```diff
@@ -0,0 +1,190 @@
+use oxc_ast::{ast::BindingIdentifier, AstKind};
+use oxc_diagnostics::{
+    miette::{self, Diagnostic},
+    thiserror::Error,
+};
+use oxc_macros::declare_oxc_lint;
+use oxc_semantic::SymbolId;
+use oxc_span::Span;
+
+use crate::{context::LintContext, rule::Rule, AstNode};
+
+#[derive(Debug, Error, Diagnostic)]
+#[error("typescript-eslint(no-unsafe-declaration-merging): Unsafe declaration merging between classes and interfaces.")]
+#[diagnostic(severity(warning))]
+struct NoUnsafeDeclarationMergingDiagnostic(#[label] pub Span);
+
+#[derive(Debug, Default, Clone)]
+pub struct NoUnsafeDeclarationMerging;
+
+declare_oxc_lint!(
+    /// ### What it does
+    ///
+    /// Disallow unsafe declaration merging.
+    ///
+    /// ### Why is this bad?
+    ///
+    /// Declaration merging between classes and interfaces is unsafe.
+    /// The TypeScript compiler doesn't check whether properties are initialized, which can cause lead to TypeScript not detecting code that will cause runtime errors.
+    ///
+    /// ### Example
+    /// ```javascript
+    /// interface Foo {}
+    /// class Foo {}
+    /// ```
+    NoUnsafeDeclarationMerging,
+    correctness
+);
+
+impl Rule for NoUnsafeDeclarationMerging {
+    fn run<'a>(&self, node: &AstNode<'a>, ctx: &LintContext<'a>) {
+        if !ctx.source_type().is_typescript() {
+            return;
+        }
+
+        let scope_symbols = ctx.semantic().scopes().get_bindings(node.scope_id());
+
+        match node.kind() {
+            AstKind::Class(decl) => {
+                if let Some(ident) = decl.id.as_ref() {
+                    for scope_symbol in scope_symbols {
+                        if let AstKind::TSInterfaceDeclaration(scope_interface) =
+                            get_symbol_kind(*scope_symbol.1, ctx)
+                        {
+                            check_and_diagnostic(ident, &scope_interface.id, decl.span, ctx);
+                        }
+                    }
+                }
+            }
+            AstKind::TSInterfaceDeclaration(decl) => {
+                for scope_symbol in scope_symbols {
+                    if let AstKind::Class(scope_class) = get_symbol_kind(*scope_symbol.1, ctx) {
+                        if let Some(scope_class_ident) = scope_class.id.as_ref() {
+                            check_and_diagnostic(&decl.id, scope_class_ident, decl.span, ctx);
+                        }
+                    }
+                }
+            }
+            _ => {}
+        }
+    }
+}
+
+fn check_and_diagnostic(
+    ident: &BindingIdentifier,
+    scope_ident: &BindingIdentifier,
+    span: Span,
+    ctx: &LintContext<'_>,
+) {
+    if scope_ident.name.as_str() == ident.name.as_str() {
+        ctx.diagnostic(NoUnsafeDeclarationMergingDiagnostic(span));
```

Comment:
Checkout crates/oxc_linter/src/rules/typescript/no_duplicate_enum_values.rs where it points to the two duplicated values, which is a great ux.

So instead of pointing to a single span, point to the two spans where they conflict.

Also, point to the name instead of the whole class / interface, because they often contain a lot of code, which will fill your terminal. We want to point to the minimal spans for the errors.